walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,116 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
 
 public final class FindMeetingQuery {
+  private static class TimeSegment {
+    public final TimeRange range;
+    public final TimeSegment left;
+    public final TimeSegment right;
```

Comment:
That would be simpler, but it would also change the steps where I search for overlapping ranges and remove them from O(log n) to O(n). :(

Reply:
Yes, premature optimization is the enemy, but also, this is literally my work for a whole week so I might as well make it as fast as possible, right?

Reply:
I could just add a really big comment that explains how it works, or try to refactor my code to make it more digestible.

Reply:
If your list is sorted, you should still be able to do a binary search on it. You're right that you'll lose out on optimizations when it comes to removing/inserting nodes. Big comments are nice. It's better if you can isolate complexity to the smallest region possible. I haven't totally digested this code yet (will do next week), but I'll give more specific feedback afterwards.

Reply:
Looked a bit closer. Pretty sure your worst case time complexity is still O(n^2) since your binary tree isn't re-balanced. Consider the case where each meeting is only 1 minute long and is progressively 2 minutes earlier than the previous meeting. Look up heaps and heap sort. They'll let you insert all your blocked out times at once in O(n log n), then you can do a single O(n) pass to merge overlapping times.

This will also let you simplify your code, since it will roughly follow the following pseudocode:
```
blockedTimes = new Heap()
for (...) {
  blockedTimes.add(event.timespan);
}

blockedTimespans = mergeTimespans(blockedTimes.toSortedArray());
availableTimespans = invert(blockedTimespans);
return firstAvailabeTimeOfSize(availableTimespans, request.rangeSize);
```

This hides all the complexity in the heap itself and in the `mergeTimespans` function.